### DISCUSSION TITLE
How meep handles two media boundaries when subpixel smoothing is turned on？

### DISCUSSION THREAD

**Discussion by Baobabs-kuang** [NanoComp/meep#2626] [Q&A]
Hello there friends from the meep community,
Once again a similar question is very disturbing to you, but I would very much like to know how to properly derive the dielectric constant information at each lattice point used by meep in the calculations.
As shown below, Figure 1 is the structure we are calculating, Figure 2 is the residual map, where the residuals are defined as shown in Figure 3, the variables labeled with _meep_ in the upper right corner are the EMF data and the dielectric constant data I got using the **_get_dft_array()_** and **_get_array()_** functions given in the meep documentation, and the residuals are supposed to be 0 everywhere, but I don't know why now at the edges of multiple media 1-2 pixels do not behave well, it seems that the data derived using the **_get_dft_array()_** and **_get_array()_** functions do not satisfy Maxwell's equations very well, which is puzzling to me.
After carefully reading meep's documentation on **_get_array()_** and **_get_dft_array()_**, I know that meep uses bilinear interpolation when dealing with geometric edges to center the EMF data and the dielectric constant data at the lattice points, but this doesn't seem to result in such large residuals from substituting the derived data into Maxwell's  equations.
I'd be very grateful if someone could address this issue I'm having and let me know how to export data that fits the Maxwell's equations at the edges. This is a critical question if you are going to use meep to help you with your topic on PINNs.
<img width="279" alt="pic1" src="https://github.com/NanoComp/meep/assets/68685880/e7073f24-15e9-4e5d-8d47-6922302b334a">
<img width="289" alt="pic2" src="https://github.com/NanoComp/meep/assets/68685880/13aba868-3546-4db0-8f44-df1c1c09c4ff">
<img width="225" alt="pic3" src="https://github.com/NanoComp/meep/assets/68685880/f5c42357-fcd7-40a7-82ee-041e27f6b509">

Here's a video of the residuals and Ez over time

https://github.com/NanoComp/meep/assets/68685880/8d9aa1f8-cc63-4070-9905-82751ecafaf2

Here's the code for simulation and visualization
```
import meep as mp
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.animation import FuncAnimation

resolution = 10         # voxels/μm
play_time = 10

n = 3.4                 # index of waveguide
w = 1                   # width of waveguide
pad = 4                 # padding between waveguide and edge of PML
dpml = 2                # thickness of PML
rad = 1.9
runtimes = np.linspace(0.01,300,100)

fig, axs = plt.subplots(1, 3, figsize=(18, 5))

def init_func():
    ani_func(-1)

def ani_func(i):
    sxy = 2*(rad+w+pad+dpml)  # cell size

    # pulse center frequency (from third-order polynomial fit)
    fcen = -0.018765*rad**3 + 0.137685*rad**2 -0.393918*rad + 0.636202
    # pulse frequency width
    df = 0.02*fcen

    src = [mp.Source(mp.GaussianSource(fcen, fwidth=df),
                        component=mp.Ez,
                        center=mp.Vector3(rad+0.1*w)),
            mp.Source(mp.GaussianSource(fcen, fwidth=df),
                        component=mp.Ez,
                        center=mp.Vector3(-(rad+0.1*w)),
                        amplitude=-1)]

    symmetries = [mp.Mirror(mp.X,phase=+1),
                    mp.Mirror(mp.Y,phase=-1)]

    geometry = [mp.Cylinder(material=mp.Medium(index=n),
                            radius=rad+w,
                            height=mp.inf,
                            center=mp.Vector3()),
                mp.Cylinder(material=mp.vacuum,
                            radius=rad,
                            height=mp.inf,
                            center=mp.Vector3())]

    calculate_area = mp.Volume(center=mp.Vector3(), size=mp.Vector3(sxy, sxy))

    sim = mp.Simulation(cell_size=mp.Vector3(sxy,sxy),
                        geometry=geometry,
                        eps_averaging=True,
                        sources=src,
                        resolution=resolution,
                        symmetries=symmetries,
                        boundary_layers=[mp.PML(dpml)])

    dft_obj = sim.add_dft_fields([mp.Ez, mp.Hx, mp.Hy], fcen, 0, 1, where=calculate_area)

    # mp.stop_when_dft_decayed() will cause the simulation stop at very early times that are not converged
    # sim.run(until=mp.stop_when_dft_decayed())

    sim.run(until=runtimes[i])

    Ez = sim.get_dft_array(dft_obj, mp.Ez, 0)
    Hx = sim.get_dft_array(dft_obj, mp.Hx, 0)
    Hy = sim.get_dft_array(dft_obj, mp.Hy, 0)
    epsilon = sim.get_array(component=mp.Dielectric, vol=calculate_area)

    DxHy = np.gradient(Hy, axis=0) * resolution
    DyHx = np.gradient(Hx, axis=1) * resolution
    omega = 2*np.pi*fcen

    residual = np.abs(DxHy - DyHx - 1j*omega*epsilon*Ez)

    for ax in axs:
        ax.cla()
    residual_img = axs[0].imshow(residual, interpolation='none', resample=False)
    axs[0].set_title(f'Residual @ runtime={runtimes[i]:.2f}')
    epsilon_img = axs[1].imshow(epsilon, interpolation='none', resample=False)
    axs[1].set_title(f'Epsilon @ runtime={runtimes[i]:.2f}')
    Ez_img = axs[2].imshow(np.abs(Ez), interpolation='none', resample=False)
    axs[2].set_title(f'Ez @ runtime={runtimes[i]:.2f}')
    if i == -1:
        plt.colorbar(residual_img, ax=axs[0])
        plt.colorbar(epsilon_img, ax=axs[1])
        plt.colorbar(Ez_img, ax=axs[2])

    sim.reset_meep()

ani = FuncAnimation(fig, ani_func, frames=len(runtimes), init_func=init_func, cache_frame_data=False)
ani.save('Residual_time.mp4', writer='ffmpeg', fps=len(runtimes)//play_time, dpi=300)

```

**Comment by stevengj** [comment]
Seems like a duplicate of #2623?
